enhancements/network/kubernetes-nmstate-operator-to-cvo.md

```diff
@@ -0,0 +1,195 @@
+---
+title: add-kubernetes-nmstate-operator-to-cvo
+authors:
+  - "@bcrochet"
+  - "@yboaron"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-03-16
+last-updated: 2021-03-16
+status: provisional
+see-also:
+  - "/enhancements/machine-config/mco-network-configuration.md"
+---
+
+# Add Kubernetes NMState Operator to CVO
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Add the Kubernetes NMState Operator to CVO.
+
+Currently, there are multiple methods of installing Kubernetes NMState into a cluster: 
+
+- CNV uses a [custom operator - CNAO](https://github.com/kubevirt/cluster-network-addons-operator#nmstate) that installs Kubernetes NMstate, among other network related items.
+The CNAO is deployed using [Hyperconverged Cluster Operator - HCO](https://github.com/kubevirt/hyperconverged-cluster-operator).
+To deploy Kubernetes NMstate it needs to deploy HCO --> CNAO --> Kubernetes NMstate
+- A [new operator](https://github.com/openshift/kubernetes-nmstate/blob/master/manifests/kubernetes-nmstate-operator.package.yaml) has been developed specifically to install Kubernetes NMstate using OLM.
+
+It is desirable for this to converge.
+
+Adding the Kubernetes NMstate as part of the CVO will allow for convenient and non-cumbersome use, which will also allow for uniformity between the various platforms (CNV, IPI-Baremetal).
```

Comment:
I think we need to be more specific here, and perhaps include some links to historical discussion, e.g:

https://github.com/openshift/enhancements/pull/161#discussion_r441112022 

IIUC the conclusion there was this shouldn't be deployed unconditionally for all platforms, and I suspect we wanted to keep it an optional add-on due to the support status of the underlying nmstate components (which is/will-be no longer the case).

Since then however we have delivered operators which target only the `baremetal` platform e.g [cluster-baremetal-operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md), so there is an established pattern of how to disable an operator on platforms where it's not relevant (although as mentioned by @russellb that's perhaps not ideal, but it's an option we can consider).

We know this is a core feature (not an optional add-on) for baremetal environments, and potentially also for vSphere and RHV?

So I think we should try to reframe this in terms of the user experience, and what is not currently possible (for example, specify secondary NIC config during the deployment phase), the CVO deploying the operator is really just an implementation detail related to achieving the desired user experience?

Reply:
+1 we need to continue that earlier discussion of "this is a core capability on (at least) bare-metal, but not on cloud". This is the core issue.